### ISSUE TITLE
Fail hard instead of pulling non-native image architecture

### ISSUE THREAD

**Issue by praiskup** [containers/podman#19717]
### Feature request description

Consider we are on ppc64le host, and pulling `docker://quay.io/centos/centos:7` -> that image doesn't provide multi-arch manifest, and `x86_64` image is silently pulled.  Is there an option that will avoid even pulling such an image?  I haven't found this in `skopeo copy` either. 

### Additional context

This has caused tricky issues with Mock for epel-7-ppc64le chroots builds when we started using --use-bootstrap-image (doing podman pull on background).

**Comment by flouthoc** [comment]
Afaik using specific `--platform` with `pull` will fail if required platform is not present on remote registry.

For example try
```
podman pull --platform linux/riscv64 alpine
```

It fails with 
```console
Trying to pull docker.io/library/alpine:latest...
Error: choosing an image from manifest list docker://alpine:latest: no image found in manifest list for architecture riscv64, variant "", OS linux
```

**Comment by vrothberg** [comment]
@flouthoc, it only errors out when hitting manifest list or OCI index.  It does not when hitting an ordinary single image manifest as `centos:7` is:
```
~ $ podman pull --platform linux/riscv64  docker://quay.io/centos/centos:7
Trying to pull quay.io/centos/centos:7...
Getting image source signatures
Copying blob 2d473b07cdd5 done   | 
Copying config 8652b9f0cb done   | 
Writing manifest to image destination
WARNING: image platform (linux/amd64) does not match the expected platform (linux/riscv64)
8652b9f0cb4c0599575e5a003f5906876e10c1ceb2ab9fe1786712dac14a50cf
```

Podman emits a warning about though when the image does not match.  Docker pulls it without error or warning.

In general, multi-arch images are a complex use case.  In an ideal world, container engines could just check and error out if the image does not match the specified architecture.  But there are too many issues to count that proved the opposite.  Some manifest lists were wrong, some image configs were wrong, there were bugs in Docker causing issues, there were bugs in Podman causing issues, there were bugs in Kaniko, buildkit ... you name it.  So, for the sake of portability, most engines do a best-effort platform matching.

For users who are willing to compromise on portability, it may be worth to think about enabling a strict platform matching.

@mtrmac WDYT?

**Comment by flouthoc** [comment]
ah i tried with docker index I guess, I think it must fail if use has specified a specific `--platform` and should WARN when no `--platform` is selected but pulled arch does not matches host arch.

**Comment by vrothberg** [comment]
> I think it must fail if user has specified a specific `--platform`

We need to look at the long history of issues preceding this one.  So far, pretty much every change to this behavior has caused regressions so we need to be extra careful and cautious.  One immediate showstopper of "must fail with --platform" is that many configs are incorrect for the above mentioned reasons.  So the image is correct but the config claims to be something different.

Hence I believe that a new flag down the lines of `--enforce-platform` may be the safest option.

**Comment by flouthoc** [comment]
Sounds fair to me. I like `--enforce-platform` and it is a safer option.

**Comment by mtrmac** [comment]
Yes, historically the platform in the config (which would be used for checking non-multi-arch images) has been incorrect several times (because nothing relied on it).

Hypothetically, it _might_ make sense to try just enforcing that again, maybe we just hit a few rare (but prominent) cases that just have been fixed. That would avoid having to maintain an option forever.

OTOH adding an option now, and having the users who need that opt in, would be a way to collect some real-world feedback without hurting most of the user base :)

**Comment by rhatdan** [comment]
Why not just put an option in containers.conf and not provide a runtime option?  Then we can run an experiment in Rawhide to change default to strict mode.

We might be approaching the state with the advent of multi-arch becoming more prominent that the cruft from years ago is being cleaned up.

And by not enforcing strict, we are breaking more users than used to be hit by bad images.

**Comment by praiskup** [comment]
Do we have a reasonably easy work-around for this?  Something that would work with RHEL 8+?

**Comment by praiskup** [comment]
Thank you for the `podman image inspect` tip, I wanted to do something like that but I haven't found a way to "assert" easily if **that** it is a "native" architecture.  In Mock we have `x86_64`, `ppc64le`, `aarch64`, ... architecture names, and we would have to have some ugly `mock -> podman` mapping.  Is there a way to query what Podman thinks is _the native architecture_?

**Comment by vrothberg** [comment]
@praiskup, you can run `podman version --format "{{.OsArch}}"`

**Comment by praiskup** [comment]
@vrothberg thanks again.

JFTR, there will be a [reusable method](https://github.com/siteshwar/openscanhub/pull/8) in Mock 5.1 to perform this assertion, if anyone else needs it.

**Comment by rhatdan** [comment]
I think this should be a breaking change for podman 5.0 to start enforcing proper arches.  Most of the problems with mixed arches are years ago, (pause container for k8s).  I think with 5.0 we should move to strict enforcement and fix users accidentally getting the wrong image.

**Comment by vrothberg** [comment]
I wholeheartedly disagree.  We tried almost every year to change _something_ on the arch checks and failed immediately.

As outlined above, I am convinced we need to make this optional (at the very least) and I'd still vote for no enforcement by default.  There is no gain when workloads break.

**Comment by rhatdan** [comment]
Should we at least warn, if we notice that the ARCH is different then the native arch, and the user did not specify it. Warning, could be disabled via containers.conf.

**Comment by vrothberg** [comment]
Yes, we do warn already when the pulled image does not match the requested platform: https://github.com/containers/common/blob/main/libimage/pull.go#L173-L181

**Comment by mheon** [comment]
This will not make 5.0. Retagging 6.0 for further discussion in that release

**Comment by mheon** [comment]
Only applies to images, not manifests. Also could break emulation.

**Comment by mheon** [comment]
If we do this as a containers.conf option, opt-in, we can do this in 6.1. Untagging 6.0
